File: apps/website/src/vault.ts
/**
 * Carnet de métadonnées.
 *
 * Le carnet ne contient jamais de mot de passe ni de clef maîtresse : seulement
 * de quoi rejouer une dérivation. Une fuite révèle les sites et les
 * identifiants, pas les mots de passe.
 *
 * Schéma : shared/vault.schema.json
 * Règles de fusion : shared/spec/vault-merge.md
 */

export const VAULT_SCHEMA = 1;
export const VAULT_STORAGE_KEY = "thecode.vault";
export const DEFAULT_LENGTH = 20;
/** Seule version admise dans le carnet : la v1 ne vit plus qu'hors carnet. */
export const VAULT_ENTRY_VERSION = 2;
export const DEFAULT_CHARSET: Charset = {
  lower: true,
  upper: true,
  symbols: true,
  numbers: true,
};

export interface Charset {
  lower: boolean;
  upper: boolean;
  symbols: boolean;
  numbers: boolean;
}

export interface VaultEntry {
  id: string;
  label?: string;
  siteKey: string;
  domains: string[];
  login?: string;
  counter: number;
  length: number;
  charset: Charset;
  v: number;
  notes?: string;
  /** Absent des entrées antérieures au champ : `updatedAt` en tient lieu. */
  createdAt?: string;
  updatedAt: string;
  deleted?: boolean;
}

export interface Vault {
  schema: number;
  updatedAt?: string;
  entries: VaultEntry[];
}

export interface Conflict {
  kind: "sitekey-divergent" | "counter-recul" | "doublon";
  entryId: string;
  detail: string;
}

function nowIso(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
}

export function emptyVault(): Vault {
  return { schema: VAULT_SCHEMA, updatedAt: nowIso(), entries: [] };
}

/** Vrai pour une entrée que le carnet accepte, c'est-à-dire en v2. */
export function isVaultEntryV2(entry: unknown): entry is VaultEntry {
  return (
    typeof entry === "object" &&
    entry !== null &&
    (entry as { v?: unknown }).v === VAULT_ENTRY_VERSION
  );
}

/**
 * Écarte toute entrée `v ≠ 2`, sans erreur.
 *
 * Appliqué à chaque lecture (stockage local, import, synchronisation) et à
 * chaque écriture : une entrée v1 qui passerait par un seul de ces chemins
 * reviendrait dans le carnet. Voir shared/spec/vault-merge.md.
 */
export function keepV2Only<T extends { entries?: unknown }>(vault: T): T {
  if (!Array.isArray(vault.entries)) return vault;
  return { ...vault, entries: vault.entries.filter(isVaultEntryV2) };
}

export function newEntry(
  siteKey: string,
  options: Partial<Omit<VaultEntry, "id" | "siteKey" | "v">> = {},
): VaultEntry {
  const now = nowIso();
  return {
    id: crypto.randomUUID(),
    label: options.label || siteKey,
    siteKey,
    domains: [...new Set(options.domains?.length ? options.domains : [siteKey])].sort(),
    login: options.login ?? "",
    counter: 1,
    length: options.length ?? DEFAULT_LENGTH,
    charset: { ...DEFAULT_CHARSET, ...(options.charset ?? {}) },
    // Toujours v2 : le carnet n'accepte pas d'autre version.
    v: VAULT_ENTRY_VERSION,
    createdAt: now,
    updatedAt: now,
  };
}

/** Entrées couvrant ce domaine. Plusieurs = plusieurs comptes sur le site. */
export function findAllByDomain(vault: Vault, domain: string): VaultEntry[] {
  const target = domain.toLowerCase();
  return vault.entries.filter(
    (e) => !e.deleted && e.domains.some((d) => d.toLowerCase() === target),
  );
}

export function findByDomain(vault: Vault, domain: string): VaultEntry | null {
  return findAllByDomain(vault, domain)[0] ?? null;
}

/**
 * Entrée d'un même site portant cet identifiant, telle que saisie.
 *
 * Un identifiant absent vaut chaîne vide : c'est ce qu'il vaut dans la
 * dérivation v2.
 */
export function findByLogin(entries: VaultEntry[], login: string): VaultEntry | null {
  return entries.find((e) => (e.login ?? "") === login) ?? null;
}

/**
 * Identifiant à proposer quand l'utilisateur n'en a saisi aucun.
 *
 * Null si un identifiant est déjà saisi, si le site est inconnu, ou si une
 * entrée sans identifiant existe : elle correspond déjà à la saisie vide.
 * Sinon, celui de la première entrée du site.
 */
export function loginToPrefill(entries: VaultEntry[], login: string): string | null {
  if (login || !entries.length || entries.some((e) => !e.login)) return null;
  return entries[0]?.login ?? null;
}

/** Entrées visibles dans l'écran de gestion : tout sauf les pierres tombales. */
export function liveEntries(vault: Vault): VaultEntry[] {
  return vault.entries.filter((e) => !e.deleted);
}

/**
 * Supprime une entrée en posant une pierre tombale.
 *
 * L'entrée reste dans le carnet, `deleted = true` et rehorodatée : effacée
 * pour de bon, l'autre appareil la ressusciterait à la synchronisation
 * suivante. Voir shared/spec/vault-merge.md.
 */
export function tombstoneEntry(vault: Vault, id: string, now = nowIso()): boolean {
  const entry = vault.entries.find((e) => e.id === id && !e.deleted);
  if (!entry) return false;
  entry.deleted = true;
  entry.updatedAt = now;
  return true;
}

/** Représentation stable, pour départager sans dépendre de l'ordre. */

/**
 * Forme canonique d'une entree, pour departager de façon déterministe.
 *
 * Compacte, clés triées a tous les niveaux, `deleted` faux retire. La forme est
 * fixée par shared/spec/vault-merge.md et non laissée a JSON.stringify : deux
 * appareils qui n'écrivent pas la même chaîne designent un gagnant différent et
 * ne convergent jamais.
 *
 * JSON.stringify avec un tableau de clés ne convient pas : il applique le
 * filtre a tous les niveaux, ce qui vidait `charset`.
 */
export function canonicalJson(entry: VaultEntry): string {
  const stable = (value: unknown): unknown => {
    if (Array.isArray(value)) return value.map(stable);
    if (value === null || typeof value !== "object") return value;
    const out: Record<string, unknown> = {};
    const record = value as Record<string, unknown>;
    for (const key of Object.keys(record).sort()) {
      if (record[key] !== undefined) out[key] = stable(record[key]);
    }
    return out;
  };

  const normalised: Record<string, unknown> = { ...entry };
  if (!normalised.deleted) delete normalised.deleted;
  return JSON.stringify(stable(normalised));
}

function mergeEntry(left: VaultEntry, right: VaultEntry, conflicts: Conflict[]): VaultEntry {
  // Égalité d'horodatage : on départage sur la représentation canonique.
  // Départager sur la position ne serait pas commutatif — chaque appareil
  // garderait le sien — et l'id ne peut pas servir, les deux entrées portent
  // la même.
  const winner =
    left.updatedAt !== right.updatedAt
      ? left.updatedAt > right.updatedAt
        ? left
        : right
      : canonicalJson(left) <= canonicalJson(right)
        ? left
        : right;

  const merged: VaultEntry = { ...winner };

  // siteKey produit le mot de passe : on ne choisit jamais à la place de
  // l'utilisateur. On garde celui de gauche et on signale.
  if (left.siteKey !== right.siteKey) {
    conflicts.push({
      kind: "sitekey-divergent",
      entryId: left.id,
      detail: `${left.siteKey} vs ${right.siteKey}`,
    });
    merged.siteKey = left.siteKey;
  }

  // Union : une addition de chaque côté ne doit pas en effacer une autre.
  merged.domains = [...new Set([...left.domains, ...right.domains])].sort();

  // Un compteur ne recule pas : une valeur haute signifie déjà renouvelé.
  const high = Math.max(left.counter, right.counter);
  const low = Math.min(left.counter, right.counter);
  merged.counter = high;
  if (high !== low && winner.counter === low) {
    conflicts.push({
      kind: "counter-recul",
      entryId: left.id,
      detail: `le plus recent porte ${low}, on garde ${high}`,
    });
  }

  // Une suppression se propage, sinon l'autre carnet ressusciterait l'entrée.
  if (left.deleted || right.deleted) merged.deleted = true;

  // Une entrée n'est créée qu'une fois : la date la plus ancienne est la
  // vraie. Un carnet antérieur au champ ne doit pas l'effacer.
  const created = [left.createdAt, right.createdAt].filter((d): d is string => Boolean(d)).sort();
  if (created.length) merged.createdAt = created[0];
  else delete merged.createdAt;

  merged.updatedAt = left.updatedAt > right.updatedAt ? left.updatedAt : right.updatedAt;
  return merged;
}

/**
 * Signale les doublons que la fusion rapproche : le même compte créé à part
 * sur deux appareils, donc sous deux id. Un doublon déjà présent d'un côté
 * l'a été quand il y est entré — le resignaler à chaque fusion serait du bruit.
 */
function findDuplicates(
  entries: VaultEntry[],
  leftIds: Set<string>,
  rightIds: Set<string>,
  conflicts: Conflict[],
) {
  const live = entries.filter((e) => !e.deleted);
  for (let i = 0; i < live.length; i += 1) {
    for (let j = i + 1; j < live.length; j += 1) {
      const a = live[i];
      const b = live[j];
      const apart =
        (leftIds.has(a.id) && !rightIds.has(a.id) && rightIds.has(b.id) && !leftIds.has(b.id)) ||
        (rightIds.has(a.id) && !leftIds.has(a.id) && leftIds.has(b.id) && !rightIds.has(b.id));
      if (!apart || (a.login ?? "") !== (b.login ?? "")) continue;
      const domains = new Set(a.domains.map((d) => d.toLowerCase()));
      if (!b.domains.some((d) => domains.has(d.toLowerCase()))) continue;
      conflicts.push({ kind: "doublon", entryId: a.id, detail: b.id });
    }
  }
}

/**
 * Fusionne deux carnets. Commutative et idempotente : l'ordre de
 * synchronisation des appareils ne doit pas changer le résultat.
 */
export function mergeVaults(left: Vault, right: Vault): { vault: Vault; conflicts: Conflict[] } {
  const conflicts: Conflict[] = [];
  const leftEntries = keepV2Only(left).entries;
  const rightEntries = keepV2Only(right).entries;
  const byId = new Map(leftEntries.map((e) => [e.id, e]));

  for (const entry of rightEntries) {
    const existing = byId.get(entry.id);
    byId.set(entry.id, existing ? mergeEntry(existing, entry, conflicts) : { ...entry });
  }

  const entries = [...byId.values()].sort((a, b) => (a.id < b.id ? -1 : a.id > b.id ? 1 : 0));
  findDuplicates(
    entries,
    new Set(leftEntries.map((e) => e.id)),
    new Set(rightEntries.map((e) => e.id)),
    conflicts,
  );
  const updatedAt =
    [left.updatedAt ?? "", right.updatedAt ?? "", ...entries.map((e) => e.updatedAt)]
      .filter(Boolean)
      .sort()
      .pop() ?? nowIso();

  return { vault: { schema: VAULT_SCHEMA, updatedAt, entries }, conflicts };
}

/**
 * Le carnet vit dans localStorage : il ne quitte pas le navigateur, comme le
 * reste du produit. Une lecture qui échoue ne doit pas casser la page — mode
 * privé, stockage bloqué — on repart d'un carnet vide.
 */
export function loadVault(): Vault {
  try {
    const raw = localStorage.getItem(VAULT_STORAGE_KEY);
    if (!raw) return emptyVault();
    const vault = JSON.parse(raw) as Vault;
    if (vault.schema !== VAULT_SCHEMA) return emptyVault();
    return keepV2Only(vault);
  } catch {
    return emptyVault();
  }
}

export function saveVault(vault: Vault): void {
  try {
    vault.updatedAt = nowIso();
    // Refus à l'écriture : une entrée v1 ne rejoint jamais le stockage.
    localStorage.setItem(VAULT_STORAGE_KEY, JSON.stringify(keepV2Only(vault)));
  } catch {
    // Stockage indisponible : le mot de passe reste dérivable, seul le
    // carnet ne persiste pas. On ne bloque pas l'utilisateur pour autant.
  }
}

/**
 * Ce qui part au serveur quand le compte a un plafond, et ce qui reste sur
 * l'appareil. Voir shared/spec/vault-sync.md, « Synchronisation partielle ».
 *
 * Ce qui est déjà sur le serveur part toujours, pierres tombales comprises :
 * une modification doit pouvoir partir. Les places libres vont aux autres
 * entrées, les plus anciennes d'abord. Une entrée jamais synchronisée puis
 * supprimée n'a rien à propager.
 */
export function selectForPush(
  vault: Vault,
  remoteIds: Iterable<string>,
  maxEntries: number,
): { push: VaultEntry[]; localOnly: VaultEntry[] } {
  const remote = new Set(remoteIds);
  const onServer = vault.entries.filter((e) => remote.has(e.id));
  const others = vault.entries
    .filter((e) => !remote.has(e.id) && !e.deleted)
    .sort((a, b) => {
      const da = a.createdAt ?? a.updatedAt;
      const db = b.createdAt ?? b.updatedAt;
      if (da !== db) return da < db ? -1 : 1;
      return a.id < b.id ? -1 : a.id > b.id ? 1 : 0;
    });

  const free = Math.max(0, maxEntries - onServer.filter((e) => !e.deleted).length);
  const byId = (a: VaultEntry, b: VaultEntry) => (a.id < b.id ? -1 : a.id > b.id ? 1 : 0);
  return {
    push: [...onServer, ...others.slice(0, free)].sort(byId),
    localOnly: others.slice(free).sort(byId),
  };
}

File: apps/cli/thecode/vault.py
"""Carnet de métadonnées TheCode.

Le carnet ne contient **jamais** de mot de passe ni de clef maîtresse : seulement
de quoi rejouer une dérivation. Une fuite révèle les sites et les identifiants,
pas les mots de passe.

Il résout trois choses que l'algorithme seul ne peut pas :

- plusieurs comptes sur un même site, via des ``siteKey`` distincts ;
- les paramètres qu'on oublie, puisqu'ils sont stockés par entrée et non
  globalement ;
- un même compte sur plusieurs domaines, via ``domains``.

Le schéma est décrit dans ``shared/vault.schema.json`` et les règles de fusion
dans ``shared/spec/vault-merge.md``.
"""

from __future__ import annotations

import json
import os
import uuid
from collections.abc import Iterable
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

SCHEMA_VERSION = 1

#: Champs jamais fusionnés automatiquement : ils déterminent le mot de passe.
_NEVER_MERGED = ("siteKey",)

#: Valeurs par défaut d'une entrée neuve, alignées sur celles des applications.
DEFAULT_CHARSET = {"lower": True, "upper": True, "symbols": True, "numbers": True}
DEFAULT_LENGTH = 20


def _canonical(entry: dict[str, Any]) -> str:
    """Forme canonique d'une entrée, pour départager de façon déterministe.

    Compacte, clés triées à tous les niveaux, ``deleted`` faux retiré. La forme
    est fixée par ``shared/spec/vault-merge.md`` et non laissée au sérialiseur
    de chaque plateforme : deux appareils qui n'écrivent pas la même chaîne
    désignent un gagnant différent et ne convergent jamais.
    """
    normalised = {k: v for k, v in entry.items() if not (k == "deleted" and not v)}
    return json.dumps(
        normalised, sort_keys=True, separators=(",", ":"), ensure_ascii=False
    )


def now_iso() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


@dataclass
class Conflict:
    """Un désaccord que la fusion refuse de trancher toute seule."""

    kind: str
    entry_id: str
    detail: str = ""


def new_entry(
    site_key: str,
    *,
    label: str = "",
    domains: Iterable[str] | None = None,
    login: str = "",
    length: int = DEFAULT_LENGTH,
    charset: dict[str, bool] | None = None,
) -> dict[str, Any]:
    """Crée une entrée.

    ``site_key`` est figé ici pour toujours : il ne suit pas les évolutions de
    la canonicalisation, sinon une mise à jour de la PSL changerait des mots de
    passe existants.

    Une entrée ne porte pas de version : elle dérive toujours en v2.
    """
    now = now_iso()
    return {
        "id": str(uuid.uuid4()),
        "label": label or site_key,
        "siteKey": site_key,
        "domains": sorted(set(domains or [site_key])),
        "login": login,
        "counter": 1,
        "length": length,
        "charset": dict(charset or DEFAULT_CHARSET),
        "createdAt": now,
        "updatedAt": now,
    }


def default_vault_path() -> Path:
    """Emplacement du carnet, suivant la convention XDG.

    Le carnet n'est pas un secret, mais il révèle les sites et identifiants :
    il mérite le dossier de configuration de l'utilisateur, pas /tmp.
    """
    base = os.environ.get("XDG_CONFIG_HOME")
    root = Path(base) if base else Path.home() / ".config"
    return root / "thecode" / "vault.json"


def empty_vault() -> dict[str, Any]:
    return {"schema": SCHEMA_VERSION, "updatedAt": now_iso(), "entries": []}


def find_by_domain(vault: dict[str, Any], domain: str) -> dict[str, Any] | None:
    """Retrouve l'entrée qui couvre ce domaine.

    C'est ce qui permet à google.com, google.fr et youtube.com de partager un
    mot de passe : ils pointent tous vers la même entrée.
    """
    target = domain.lower()
    for entry in vault.get("entries", []):
        if entry.get("deleted"):
            continue
        if target in (d.lower() for d in entry.get("domains", [])):
            return entry
    return None


def find_all_by_domain(vault: dict[str, Any], domain: str) -> list[dict[str, Any]]:
    """Toutes les entrées couvrant ce domaine : plusieurs comptes sur un site."""
    target = domain.lower()
    return [
        e
        for e in vault.get("entries", [])
        if not e.get("deleted") and target in (d.lower() for d in e.get("domains", []))
    ]


def _merge_entry(
    left: dict[str, Any], right: dict[str, Any], conflicts: list[Conflict]
) -> dict[str, Any]:
    """Fusionne deux versions d'une même entrée. Voir shared/spec/vault-merge.md."""
    # Égalité d'horodatage : on départage sur la représentation canonique, la
    # plus petite l'emportant. Départager sur la position ne serait pas
    # commutatif — chaque appareil garderait le sien — et l'id ne peut pas
    # servir puisque les deux entrées portent la même.
    if left["updatedAt"] != right["updatedAt"]:
        winner = left if left["updatedAt"] > right["updatedAt"] else right
    else:
        winner = min(left, right, key=_canonical)

    loser = right if winner is left else left

    merged = dict(winner)

    # siteKey produit le mot de passe : on ne choisit jamais à la place de
    # l'utilisateur. On garde celui de gauche et on signale.
    for key in _NEVER_MERGED:
        if left.get(key) != right.get(key):
            conflicts.append(
                Conflict(
                    "sitekey-divergent",
                    left["id"],
                    f"{left.get(key)!r} vs {right.get(key)!r}",
                )
            )
            merged[key] = left[key]

    # Union : une addition de chaque côté ne doit pas en effacer une autre.
    merged["domains"] = sorted(set(left.get("domains", [])) | set(right.get("domains", [])))

    # Un compteur ne recule pas : une valeur haute signifie déjà renouvelé.
    high, low = max(left["counter"], right["counter"]), min(left["counter"], right["counter"])
    merged["counter"] = high
    if high != low and winner["counter"] == low:
        conflicts.append(
            Conflict(
                "counter-recul",
                left["id"],
                f"le plus recent porte {low}, on garde {high}",
            )
        )

    # Une suppression se propage, sinon l'autre carnet ressusciterait l'entrée.
    if left.get("deleted") or right.get("deleted"):
        merged["deleted"] = True

    # Une entrée n'est créée qu'une fois : la date la plus ancienne est la
    # vraie. Un carnet antérieur au champ ne doit pas l'effacer.
    created = [d for d in (left.get("createdAt"), right.get("createdAt")) if d]
    if created:
        merged["createdAt"] = min(created)
    else:
        merged.pop("createdAt", None)

    merged["updatedAt"] = max(left["updatedAt"], right["updatedAt"])
    _ = loser
    return merged


def _find_duplicates(
    entries: list[dict[str, Any]],
    left_ids: set[str],
    right_ids: set[str],
    conflicts: list[Conflict],
) -> None:
    """Signale les doublons que la fusion rapproche.

    Le même compte créé à part sur deux appareils porte deux id. Un doublon
    déjà présent d'un côté l'a été quand il y est entré — le resignaler à
    chaque fusion serait du bruit.
    """
    live = [e for e in entries if not e.get("deleted")]
    for i, a in enumerate(live):
        for b in live[i + 1 :]:
            apart = (
                a["id"] in left_ids
                and a["id"] not in right_ids
                and b["id"] in right_ids
                and b["id"] not in left_ids
            ) or (
                a["id"] in right_ids
                and a["id"] not in left_ids
                and b["id"] in left_ids
                and b["id"] not in right_ids
            )
            if not apart or (a.get("login") or "") != (b.get("login") or ""):
                continue
            domains = {d.lower() for d in a.get("domains", [])}
            if not any(d.lower() in domains for d in b.get("domains", [])):
                continue
            conflicts.append(Conflict("doublon", a["id"], b["id"]))


def merge(
    left: dict[str, Any], right: dict[str, Any]
) -> tuple[dict[str, Any], list[Conflict]]:
    """Fusionne deux carnets.

    Commutative et idempotente : l'ordre de synchronisation des appareils ne
    doit pas changer le résultat, sinon ils ne convergent jamais.
    """
    conflicts: list[Conflict] = []
    by_id: dict[str, dict[str, Any]] = {e["id"]: e for e in left.get("entries", [])}

    for entry in right.get("entries", []):
        existing = by_id.get(entry["id"])
        by_id[entry["id"]] = (
            _merge_entry(existing, entry, conflicts) if existing else dict(entry)
        )

    entries = sorted(by_id.values(), key=lambda e: e["id"])
    _find_duplicates(
        entries,
        {e["id"] for e in left.get("entries", [])},
        {e["id"] for e in right.get("entries", [])},
        conflicts,
    )
    updated = max(
        [left.get("updatedAt", ""), right.get("updatedAt", "")]
        + [e["updatedAt"] for e in entries],
        default=now_iso(),
    )
    return {"schema": SCHEMA_VERSION, "updatedAt": updated, "entries": entries}, conflicts


def select_for_push(
    vault: dict[str, Any], remote_ids: Iterable[str], max_entries: int
) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
    """Ce qui part au serveur quand le compte a un plafond, et ce qui reste.

    Voir shared/spec/vault-sync.md, « Synchronisation partielle ». Ce qui est
    déjà sur le serveur part toujours, pierres tombales comprises : une
    modification doit pouvoir partir. Les places libres vont aux autres
    entrées, les plus anciennes d'abord. Une entrée jamais synchronisée puis
    supprimée n'a rien à propager.
    """
    remote = set(remote_ids)
    entries = vault.get("entries", [])
    on_server = [e for e in entries if e["id"] in remote]
    others = sorted(
        (e for e in entries if e["id"] not in remote and not e.get("deleted")),
        key=lambda e: (e.get("createdAt") or e["updatedAt"], e["id"]),
    )

    free = max(0, max_entries - len([e for e in on_server if not e.get("deleted")]))
    push = sorted(on_server + others[:free], key=lambda e: e["id"])
    local_only = sorted(others[free:], key=lambda e: e["id"])
    return push, local_only


def load(path: Path) -> dict[str, Any]:
    if not path.is_file():
        return empty_vault()
    data = json.loads(path.read_text(encoding="utf-8"))
    if data.get("schema") != SCHEMA_VERSION:
        raise ValueError(
            f"Carnet en version {data.get('schema')}, attendu {SCHEMA_VERSION}"
        )
    return data


def save(vault: dict[str, Any], path: Path) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    vault["updatedAt"] = now_iso()
    # Écriture atomique : une interruption ne doit pas laisser un carnet
    # tronqué, qui ferait perdre toutes les entrées.
    tmp = path.with_suffix(path.suffix + ".tmp")
    tmp.write_text(json.dumps(vault, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
    tmp.replace(path)
